File: go/luhn/luhn.go
package luhn

//Valid determine whether a provided string is a valid
func Valid(s string) bool {
	realPos := 0
	sum := 0
	for pos := len(s) - 1; pos >= 0; pos-- {
		if s[pos] == ' ' {
			continue
		}
		if s[pos] < '0' || s[pos] > '9' {
			return false
		}
		val := int(s[pos] - '0')
		if realPos%2 == 1 {
			val *= 2
			if val > 9 {
				val -= 9
			}
		}
		realPos++
		sum += val
	}
	return sum%10 == 0 && realPos > 1
}

File: python/luhn/luhn.py
import typing
import itertools
from string import digits

class Luhn:
    def __init__(self, card_num: str) -> None:
        self.num = card_num.replace(' ','')


    '''for fun
    def valid(self) -> bool:
        """Confirm whether or not the provided card number is a valid number"""
        return (self.num.isdigit()) and len(self.num) != 1 and\
                (sum(itertools.chain([int(val) for val in self.num[::-2]],
                 [(dub := int(val) * 2) - (0 if dub <= 9 else 9) for val in self.num[-2::-2]])) % 10) == 0
    '''

    # for readability
    def valid(self) -> bool:
        """Confirm whether or not the provided card number is a valid number"""
        if not self.num.isdigit() or len(self.num) == 1:
            return False
        same_vals = [int(i) for i in self.num[::-2]]
        doubled = [int(val) * 2 for val in self.num[-2::-2]]
        combined = [num - (0 if num <= 9 else 9) for num in doubled]
        combined.extend(same_vals)
        return (sum(combined) % 10) == 0